File: a105_mpu6500_spi_driver/src/mpu6500_spi_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/imu.hpp>

#include <linux/spi/spidev.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <unistd.h>
#include <cstring>
#include <cmath>
#include <vector>
#include <string>

namespace {
constexpr uint8_t REG_SMPLRT_DIV     = 0x19;
constexpr uint8_t REG_CONFIG         = 0x1A;
constexpr uint8_t REG_GYRO_CONFIG    = 0x1B;
constexpr uint8_t REG_ACCEL_CONFIG   = 0x1C;
constexpr uint8_t REG_ACCEL_CONFIG2  = 0x1D;
constexpr uint8_t REG_INT_PIN_CFG    = 0x37;
constexpr uint8_t REG_INT_ENABLE     = 0x38;
constexpr uint8_t REG_ACCEL_XOUT_H   = 0x3B; // 14 bytes from here
constexpr uint8_t REG_USER_CTRL      = 0x6A;
constexpr uint8_t REG_PWR_MGMT_1     = 0x6B;
constexpr uint8_t REG_WHO_AM_I       = 0x75;

inline int16_t be16(const uint8_t *p) {
  return static_cast<int16_t>((p[0] << 8) | p[1]);
}
}

class Mpu6500SpiNode : public rclcpp::Node {
public:
  Mpu6500SpiNode() : Node("mpu6500_spi_node") {
    device_         = this->declare_parameter<std::string>("device", "/dev/spidev0.0");
    frame_id_       = this->declare_parameter<std::string>("frame_id", "imu_link");
    rate_hz_        = this->declare_parameter<int>("rate_hz", 200);
    spi_speed_hz_   = this->declare_parameter<int>("spi_speed_hz", 1000000); // 1 MHz
    gyro_dps_       = this->declare_parameter<int>("gyro_range_dps", 2000);  // 250|500|1000|2000
    accel_g_        = this->declare_parameter<int>("accel_range_g", 16);     // 2|4|8|16
    dlpf_cfg_       = this->declare_parameter<int>("dlpf_cfg", 3);  // 0..6
    smplrt_div_     = this->declare_parameter<int>("smplrt_div", 4);
    double gav = this->declare_parameter<double>("gyro_variance", 0.02);
    double aav = this->declare_parameter<double>("accel_variance", 0.04);

    if (!open_spi()) {
      throw std::runtime_error("SPI open/config failed");
    }
    if (!init_mpu()) {
      throw std::runtime_error("MPU init failed");
    }

    imu_pub_ = this->create_publisher<sensor_msgs::msg::Imu>("imu/data_raw", rclcpp::SensorDataQoS());

    // set covariances
    gyro_cov_.fill(0.0);
    accel_cov_.fill(0.0);
    gyro_cov_[0] = gyro_cov_[4] = gyro_cov_[8] = gav;
    accel_cov_[0] = accel_cov_[4] = accel_cov_[8] = aav;

    timer_ = this->create_wall_timer(
        std::chrono::milliseconds(1000 / std::max(1, rate_hz_)),
        std::bind(&Mpu6500SpiNode::poll_and_publish, this));
  }

  ~Mpu6500SpiNode() override {
    if (fd_ >= 0) ::close(fd_);
  }

private:
  bool open_spi() {
    fd_ = ::open(device_.c_str(), O_RDWR);
    if (fd_ < 0) {
      RCLCPP_ERROR(get_logger(), "open(%s) failed", device_.c_str());
      return false;
    }
    uint8_t mode = SPI_MODE_0;
    uint8_t bits = 8;
    if (ioctl(fd_, SPI_IOC_WR_MODE, &mode) < 0) return false;
    if (ioctl(fd_, SPI_IOC_WR_BITS_PER_WORD, &bits) < 0) return false;
    if (ioctl(fd_, SPI_IOC_WR_MAX_SPEED_HZ, &spi_speed_hz_) < 0) return false;
    return true;
  }

  bool write_reg(uint8_t reg, uint8_t val) {
    uint8_t tx[2] = { static_cast<uint8_t>(reg & 0x7F), val };
    struct spi_ioc_transfer tr{};
    tr.tx_buf = reinterpret_cast<unsigned long>(tx);
    tr.len = 2;
    tr.speed_hz = spi_speed_hz_;
    tr.bits_per_word = 8;
    return ioctl(fd_, SPI_IOC_MESSAGE(1), &tr) >= 1;
  }

  bool read_regs(uint8_t reg, uint8_t *dst, size_t n) {
    std::vector<uint8_t> tx(n + 1, 0x00), rx(n + 1, 0x00);
    tx[0] = reg | 0x80; // read
    struct spi_ioc_transfer tr{};
    tr.tx_buf = reinterpret_cast<unsigned long>(tx.data());
    tr.rx_buf = reinterpret_cast<unsigned long>(rx.data());
    tr.len = tx.size();
    tr.speed_hz = spi_speed_hz_;
    tr.bits_per_word = 8;
    if (ioctl(fd_, SPI_IOC_MESSAGE(1), &tr) < 1) return false;
    std::memcpy(dst, rx.data() + 1, n);
    return true;
  }

  bool init_mpu() {
    // reset
    write_reg(REG_PWR_MGMT_1, 0x80);
    rclcpp::sleep_for(std::chrono::milliseconds(100));
    // clock: PLL with X axis gyro
    if (!write_reg(REG_PWR_MGMT_1, 0x01)) return false;
    // disable I2C interface for SPI
    if (!write_reg(REG_USER_CTRL, 0x10)) return false;

    // WHO_AM_I check: 0x70 (MPU6500) or 0x71 (MPU9250)
    uint8_t who = 0;
    if (!read_regs(REG_WHO_AM_I, &who, 1)) return false;
    if (!(who == 0x70 || who == 0x71)) {
      RCLCPP_WARN(get_logger(), "WHO_AM_I=0x%02X not expected", who);
    }

    // DLPF
    write_reg(REG_CONFIG, dlpf_cfg_ & 0x07);
    // sample rate = gyro_rate/(1+SMPLRT_DIV). With DLPF on, gyro_rate=1kHz.
    write_reg(REG_SMPLRT_DIV, smplrt_div_ & 0xFF);

    // gyro range
    uint8_t gcfg = 0;
    if      (gyro_dps_ == 250)  gcfg = 0 << 3;
    else if (gyro_dps_ == 500)  gcfg = 1 << 3;
    else if (gyro_dps_ == 1000) gcfg = 2 << 3;
    else                        gcfg = 3 << 3; // 2000
    write_reg(REG_GYRO_CONFIG, gcfg);

    // accel range
    uint8_t acfg = 0;
    if      (accel_g_ == 2)  acfg = 0 << 3;
    else if (accel_g_ == 4)  acfg = 1 << 3;
    else if (accel_g_ == 8)  acfg = 2 << 3;
    else                     acfg = 3 << 3; // 16 g
    write_reg(REG_ACCEL_CONFIG, acfg);

    write_reg(REG_ACCEL_CONFIG2, (dlpf_cfg_ & 0x07));

    // latched interrupts off, data ready off
    write_reg(REG_INT_PIN_CFG, 0x00);
    write_reg(REG_INT_ENABLE,  0x00);
    compute_scales();
    return true;
  }

  void compute_scales() {
    // LSB/FS
    switch (accel_g_) {
      case 2:  accel_lsb_per_g_ = 16384.0; break;
      case 4:  accel_lsb_per_g_ = 8192.0;  break;
      case 8:  accel_lsb_per_g_ = 4096.0;  break;
      default: accel_lsb_per_g_ = 2048.0;  break; // 16g
    }
    switch (gyro_dps_) {
      case 250:  gyro_lsb_per_dps_ = 131.0;  break;
      case 500:  gyro_lsb_per_dps_ = 65.5;   break;
      case 1000: gyro_lsb_per_dps_ = 32.8;   break;
      default:   gyro_lsb_per_dps_ = 16.4;   break; // 2000
    }
  }

  void poll_and_publish() {
    uint8_t buf[14];
    if (!read_regs(REG_ACCEL_XOUT_H, buf, 14)) {
      RCLCPP_WARN_THROTTLE(get_logger(), *this->get_clock(), 5000, "read failed");
      return;
    }

    int16_t ax = be16(buf + 0);
    int16_t ay = be16(buf + 2);
    int16_t az = be16(buf + 4);
    int16_t gx = be16(buf + 8);
    int16_t gy = be16(buf + 10);
    int16_t gz = be16(buf + 12);

    sensor_msgs::msg::Imu msg;
    msg.header.stamp = this->now();
    msg.header.frame_id = frame_id_;

    // accel in m/s^2
    constexpr double G = 9.80665;
    msg.linear_acceleration.x = static_cast<double>(ax) / accel_lsb_per_g_ * G;
    msg.linear_acceleration.y = static_cast<double>(ay) / accel_lsb_per_g_ * G;
    msg.linear_acceleration.z = static_cast<double>(az) / accel_lsb_per_g_ * G;

    // gyro in rad/s
    constexpr double DEG2RAD = M_PI / 180.0;
    msg.angular_velocity.x = (static_cast<double>(gx) / gyro_lsb_per_dps_) * DEG2RAD;
    msg.angular_velocity.y = (static_cast<double>(gy) / gyro_lsb_per_dps_) * DEG2RAD;
    msg.angular_velocity.z = (static_cast<double>(gz) / gyro_lsb_per_dps_) * DEG2RAD;

    // orientation unknown
    msg.orientation_covariance[0] = -1.0;
    // covariances
    for (int i = 0; i < 9; ++i) {
      msg.angular_velocity_covariance[i] = gyro_cov_[i];
      msg.linear_acceleration_covariance[i] = accel_cov_[i];
    }

    imu_pub_->publish(std::move(msg));
  }

  // params
  std::string device_, frame_id_;
  int rate_hz_, spi_speed_hz_, gyro_dps_, accel_g_, dlpf_cfg_, smplrt_div_;
  double accel_lsb_per_g_{16384.0}, gyro_lsb_per_dps_{16.4};
  std::array<double,9> gyro_cov_{}, accel_cov_{};

  // io
  int fd_{-1};

  // ros
  rclcpp::Publisher<sensor_msgs::msg::Imu>::SharedPtr imu_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  try {
    rclcpp::spin(std::make_shared<Mpu6500SpiNode>());
  } catch (const std::exception &e) {
    fprintf(stderr, "fatal: %s\n", e.what());
  }
  rclcpp::shutdown();
  return 0;
}

File: a105_mpu6500_spi_py/a105_mpu6500_spi_py/mpu6500_spi_node.py
#!/usr/bin/env python3
import math, time, spidev
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Imu

REG_SMPLRT_DIV, REG_CONFIG, REG_GYRO_CONFIG, REG_ACCEL_CONFIG = 0x19,0x1A,0x1B,0x1C
REG_ACCEL_CONFIG2, REG_INT_PIN_CFG, REG_INT_ENABLE = 0x1D,0x37,0x38
REG_ACCEL_XOUT_H, REG_USER_CTRL, REG_PWR_MGMT_1, REG_WHO_AM_I = 0x3B,0x6A,0x6B,0x75

G = 9.80665
DEG2RAD = math.pi/180.0

def be16(hi, lo):
    v = (hi<<8)|lo
    return v-65536 if v & 0x8000 else v

class Mpu6500Node(Node):
    def __init__(self):
        super().__init__('mpu6500_spi_node')

        self.declare_parameter('bus', 0)
        self.declare_parameter('dev', 0)
        self.declare_parameter('speed', 1_000_000)
        self.declare_parameter('rate', 200.0)
        self.declare_parameter('gyro_range', 2000)
        self.declare_parameter('accel_range', 16)
        self.declare_parameter('dlpf', 3)
        self.declare_parameter('smplrt_div', 4)
        self.declare_parameter('frame_id', 'imu_link')
        self.declare_parameter('gyro_variance', 0.02)
        self.declare_parameter('accel_variance', 0.04)

        bus         = int(self.get_parameter('bus').value)
        dev         = int(self.get_parameter('dev').value)
        speed       = int(self.get_parameter('speed').value)
        self.rate   = float(self.get_parameter('rate').value)
        self.frame  = str(self.get_parameter('frame_id').value)
        self.gvar   = float(self.get_parameter('gyro_variance').value)
        self.avar   = float(self.get_parameter('accel_variance').value)

        gyro_dps    = int(self.get_parameter('gyro_range').value)
        accel_g     = int(self.get_parameter('accel_range').value)
        dlpf        = int(self.get_parameter('dlpf').value)
        smpldiv     = int(self.get_parameter('smplrt_div').value)

        self.acc_lsb = {2:16384.0, 4:8192.0, 8:4096.0}.get(accel_g, 2048.0)
        self.gyr_lsb = {250:131.0, 500:65.5, 1000:32.8}.get(gyro_dps, 16.4)

        self.spi = spidev.SpiDev()
        self.spi.open(bus, dev)         # /dev/spidev0.0
        self.spi.mode = 0
        self.spi.max_speed_hz = speed
        self.spi.bits_per_word = 8

        self._init_device(gyro_dps, accel_g, dlpf, smpldiv)

        self.pub = self.create_publisher(Imu, 'imu/data_raw', qos_profile_sensor_data)
        period = 1.0/max(self.rate, 1e-3)
        self.timer = self.create_timer(period, self._tick)

    def _w(self, reg, val): self.spi.xfer2([reg & 0x7F, val & 0xFF])
    def _rn(self, reg, n):  return self.spi.xfer2([reg | 0x80] + [0]*n)[1:]

    def _init_device(self, gyro_dps, accel_g, dlpf, smpldiv):
        self._w(REG_PWR_MGMT_1, 0x80); time.sleep(0.1)
        self._w(REG_PWR_MGMT_1, 0x01)
        self._w(REG_USER_CTRL, 0x10)
        who = self._rn(REG_WHO_AM_I, 1)[0]
        if who not in (0x70, 0x71):
            self.get_logger().warn(f'WHO_AM_I=0x{who:02X}')

        self._w(REG_CONFIG, dlpf & 0x07)
        self._w(REG_SMPLRT_DIV, smpldiv & 0xFF)
        gcfg = {250:0,500:1,1000:2}.get(gyro_dps,3) << 3
        self._w(REG_GYRO_CONFIG, gcfg)
        acfg = {2:0,4:1,8:2}.get(accel_g,3) << 3
        self._w(REG_ACCEL_CONFIG, acfg)
        self._w(REG_ACCEL_CONFIG2, dlpf & 0x07)
        self._w(REG_INT_PIN_CFG, 0x00); self._w(REG_INT_ENABLE, 0x00)

    def _tick(self):
        try:
            b = self._rn(REG_ACCEL_XOUT_H, 14)
        except Exception as e:
            self.get_logger().throttle(5.0).error(f'spi read failed: {e}')
            return

        ax = be16(b[0],b[1]); ay = be16(b[2],b[3]); az = be16(b[4],b[5])
        gx = be16(b[8],b[9]); gy = be16(b[10],b[11]); gz = be16(b[12],b[13])

        msg = Imu()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self.frame

        msg.linear_acceleration.x = ax/self.acc_lsb*G
        msg.linear_acceleration.y = ay/self.acc_lsb*G
        msg.linear_acceleration.z = az/self.acc_lsb*G


        msg.angular_velocity.x = (gx/self.gyr_lsb)*DEG2RAD
        msg.angular_velocity.y = (gy/self.gyr_lsb)*DEG2RAD
        msg.angular_velocity.z = (gz/self.gyr_lsb)*DEG2RAD

        msg.orientation_covariance[0] = -1.0
        for i in (0,4,8):
            msg.angular_velocity_covariance[i] = self.gvar
            msg.linear_acceleration_covariance[i] = self.avar

        self.pub.publish(msg)

    def destroy_node(self):
        try: self.spi.close()
        except Exception: pass
        super().destroy_node()

def main():
    rclpy.init()
    node = Mpu6500Node()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()
